boundary.go

```diff
@@ -43,42 +50,81 @@ func (b *boundaryReader) Read(dest []byte) (n int, err error) {
 		return b.buffer.Read(dest)
 	}
 
-	peek, err := b.r.Peek(peekBufferSize)
-	peekEOF := (err == io.EOF)
-	if err != nil && !peekEOF && err != bufio.ErrBufferFull {
-		// Unexpected error
-		return 0, errors.WithStack(err)
-	}
-	var nCopy int
-	idx, complete := locateBoundary(peek, b.nlPrefix)
-	if idx != -1 {
-		// Peeked boundary prefix, read until that point
-		nCopy = idx
-		if !complete && nCopy == 0 {
-			// Incomplete boundary, move past it
-			nCopy = 1
+	for i := 0; i < cap(dest); i++ {
+		oneByte, err := b.r.Peek(1)
+		if err != nil && err != io.EOF {
+			return 0, errors.WithStack(err)
 		}
-	} else {
-		// No boundary found, move forward a safe distance
-		if nCopy = len(peek) - len(b.nlPrefix) - 1; nCopy <= 0 {
-			nCopy = 0
-			if peekEOF {
-				// No more peek space remaining and no boundary found
-				return 0, errors.WithStack(io.ErrUnexpectedEOF)
+		if bytes.Equal(oneByte, []byte{'\n'}) {
+			peek, err := b.r.Peek(len(b.nlPrefix) + 2)
+			switch err {
+			case nil:
+				// don't check the boundary until we only have one new line
+				if bytes.HasPrefix(peek, []byte("\n\n")) ||
+					bytes.HasPrefix(peek, []byte("\n\r")) {
+					break
+				}
+				// check for a match against boundary or terminal boundary
+				if b.isDelimiter(peek[1:]) || b.isTerminator(peek[1:]) {
+					// if we stashed a carriage return, lets pop that back onto the io.Reader
+					if b.crBoundaryPrefix {
+						err = b.r.UnreadByte()
+						switch err {
+						case nil, bufio.ErrInvalidUnreadByte:
+							// carriage return boundary prefix bit already unread
+						default:
+							// this should never happen, fatal error
+							return 0, errors.WithStack(err)
+						}
+						b.crBoundaryPrefix = false
+					}
+					n, err = b.buffer.Read(dest)
+					switch err {
+					case nil, io.EOF:
+						return n, io.EOF
+					default:
+						return 0, errors.WithStack(err)
+					}
+				}
+			default:
+				// got to the end without seeing a boundary
+				if err == io.EOF {
```

Comment:
I don't understand why we are in a `switch err` but then testing err for equality, can we pull this out into a `io.EOF:` block?